### PR TITLE
Move row expression to spi

### PR DESCRIPTION
**Pull request by hellium01**
This is part of #12375 and on top of #12472

### CODE REVIEW

File: presto-spi/src/main/java/com/facebook/presto/spi/relation/row/VariableReferenceExpression.java [hunk 1]
```diff
@@ -11,7 +11,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package com.facebook.presto.sql.relational;
+package com.facebook.presto.spi.relation.row;
```
Comment (highker) [inline presto-spi/src/main/java/com/facebook/presto/spi/relation/row/VariableReferenceExpression.java]: package `spi.relation` is good enough?

File: presto-spi/src/main/java/com/facebook/presto/spi/type/FunctionType.java [hunk 2]
```diff
@@ -71,7 +68,11 @@ public Type getReturnType()
     @Override
     public List<Type> getTypeParameters()
     {
-        return ImmutableList.<Type>builder().addAll(argumentTypes).add(returnType).build();
+        List<Type> parameters = new ArrayList<>(argumentTypes.size() + 1);
+        argumentTypes.stream()
```
Comment (highker) [inline presto-spi/src/main/java/com/facebook/presto/spi/type/FunctionType.java]: same line

File: presto-spi/src/main/java/com/facebook/presto/spi/relation/row/VariableReferenceExpression.java (hunk 1 above)
Comment (hellium01) [reply presto-spi/src/main/java/com/facebook/presto/spi/relation/row/VariableReferenceExpression.java]: OK, forget we are not going to add table level expression but convert planNode instead...

File: presto-main/src/test/java/com/facebook/presto/sql/TestRowExpressionSerde.java [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+    {
+        RowExpression roundTrip = getRoudTrip(sql);
+        assertTrue(roundTrip instanceof ConstantExpression);
+        String roundTripValue = ((Slice) ((ConstantExpression) roundTrip).getValue()).toStringUtf8();
+        Type roundTripType = roundTrip.getType();
+        assertEquals(roundTripValue, expectedString);
+        assertEquals(roundTripType, expectedType);
+    }
+
+    private RowExpression getRoudTrip(String sql)
```
Comment (highker) [inline presto-main/src/test/java/com/facebook/presto/sql/TestRowExpressionSerde.java]: typo `getRoudTrip`

File: presto-main/src/test/java/com/facebook/presto/sql/TestRowExpressionSerde.java [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+    {
+        RowExpression rowExpression = translate(expression(sql, new ParsingOptions(AS_DOUBLE)));
+        String json = codec.toJson(rowExpression);
+        return codec.fromJson(json);
+    }
+
+    private JsonCodec<RowExpression> getJsonCodec()
+            throws Exception
+    {
+        Module module = new Module()
```
Comment (highker) [inline presto-main/src/test/java/com/facebook/presto/sql/TestRowExpressionSerde.java]: Use lambda
```java
Module module = binder -> {
    binder.install(new JsonModule());
    configBinder(binder).bindConfig(FeaturesConfig.class);

    binder.bind(TypeManager.class).to(TypeRegistry.class).in(Scopes.SINGLETON);
    jsonBinder(binder).addDeserializerBinding(Type.class).to(TypeDeserializer.class);
    newSetBinder(binder, Type.class);

    binder.bind(BlockEncodingSerde.class).to(BlockEncodingManager.class).in(Scopes.SINGLETON);
    newSetBinder(binder, BlockEncoding.class);
    jsonBinder(binder).addSerializerBinding(Block.class).to(BlockJsonSerde.Serializer.class);
    jsonBinder(binder).addDeserializerBinding(Block.class).to(BlockJsonSerde.Deserializer.class);
    jsonCodecBinder(binder).bindJsonCodec(RowExpression.class);
};

File: presto-main/src/test/java/com/facebook/presto/sql/TestRowExpressionSerde.java [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+        {
+            @Override
+            public void configure(Binder binder)
+            {
+                binder.install(new JsonModule());
+                configBinder(binder).bindConfig(FeaturesConfig.class);
+
+                binder.bind(TypeRegistry.class).in(Scopes.SINGLETON);
```
Comment (highker) [inline presto-main/src/test/java/com/facebook/presto/sql/TestRowExpressionSerde.java]: this line is useless

File: presto-main/src/test/java/com/facebook/presto/sql/TestRowExpressionSerde.java [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+                binder.bind(TypeManager.class).to(TypeRegistry.class).in(Scopes.SINGLETON);
+                jsonBinder(binder).addDeserializerBinding(Type.class).to(TypeDeserializer.class);
+                newSetBinder(binder, Type.class);
+
+                binder.bind(BlockEncodingManager.class).in(Scopes.SINGLETON);
```
Comment (highker) [inline presto-main/src/test/java/com/facebook/presto/sql/TestRowExpressionSerde.java]: this line is useless

**Comment by hellium01** [comment]
Let's talk tomorrow. I added test case for array subscript with integer and it seems to be OK.

**Comment by highker** [comment]
Try `"CAST(ROW(1) AS ROW(col1 integer)).col1"`; it will fail

**Comment by hellium01** [comment]
I will address that in other PR. This is a result of hand written constant that has arbitrary type does not agree with presto Types's java type.

**Pull request merged**